holoviews/operation/datashader.py

```diff
@@ -1725,3 +1725,193 @@ class directly_connect_edges(_connect_edges, connect_edges):
+        if np.isnan(val):
+            val = self.p.null_value
+
+        if ((self.p.value_bounds and
+             not (self.p.value_bounds[0] < val < self.p.value_bounds[1]))
+             or val == self.p.null_value):
+            empty_df = self._empty_df(raster.dataset)
+            df = self.p.points_transformer(self.p.transformer(empty_df))
+            self.hits = self.p.hits_transformer(self.p.transformer(empty_df))
+            vdims = self._vdims(raster, df)
+            return Points(df, kdims=raster.kdims, vdims=vdims)
+
+        masked = self._mask_dataframe(raster, x, y, xdelta, ydelta)
+        dist_sorted = self._sort_by_distance(raster, masked, x, y)
+        df = self.p.points_transformer(self.p.transformer(dist_sorted))
+        self.hits = self.p.hits_transformer(self.p.transformer(dist_sorted))
+        vdims = self._vdims(raster, df)
+        return Points(df, kdims=raster.kdims, vdims=vdims).iloc[:self.p.point_count]
+
+    @property
+    def mask(self):
+        return inspect_mask.instance(pixels=self.p.pixels)
+
+    @classmethod
+    def _vdims(cls, raster, df):
+        ds = raster.dataset
+        if 'spatialpandas' in ds.interface.datatype:
+            coords = [ds.interface.geo_column(ds.data)]
+        else:
+            coords = [kd.name for kd in raster.kdims]
+        return [col for col in df.columns if col not in coords]
+
+    @classmethod
+    def _empty_df(cls, dataset):
+        if 'dask' in dataset.interface.datatype:
+            return dataset.data._meta.iloc[:0]
+        elif dataset.interface.datatype in ['pandas', 'geopandas', 'spatialpandas']:
+            return dataset.data.head(0)
+        return dataset.iloc[:0].dframe()
+
+    @classmethod
+    def _distance_args(cls, element, x_range, y_range,  pixels):
+        ycount, xcount =  element.interface.shape(element, gridded=True)
+        x_delta = abs(x_range[1] - x_range[0]) / xcount
+        y_delta = abs(y_range[1] - y_range[0]) / ycount
+        return (x_delta*pixels, y_delta*pixels)
+
+    @classmethod
+    def _mask_dataframe(cls, raster, x, y, xdelta, ydelta):
+        "Mask the dataframe with size around x and y"
+        ds = raster.dataset
+        x0, x1, y0, y1 = x-xdelta, x+xdelta, y-ydelta, y+ydelta
+        if 'spatialpandas' in ds.interface.datatype:
+            df = ds.data.cx[x0:x1, y0:y1]
+            return df.compute() if hasattr(df, 'compute') else df
+        xdim, ydim = raster.kdims
+        query = {xdim.name: (x0, x1), ydim.name: (y0, y1)}
+        return ds.select(**query).dframe()
+
+    @classmethod
+    def _get_xs_ys(cls, raster, df):
+        "Returns the x and y arrays or series for the given dataframe"
+        ds = raster.dataset
+        if 'spatialpandas' in ds.interface.datatype:
+            geom = ds.interface.geo_column(ds.data)
+            return df[geom].array.x, df[geom].array.y
+        xdim, ydim = raster.kdims
+        return df[xdim].values, df[ydim].values
```

Comment:
It might be good to make this a module-level function; I've needed precisely this functionality to handle spatialpandas and regular pandas dataframes consistently.

Reply:
I'm happy to pull out utilities but I can't quite see it here. What would be the name of the utility and what would be its signature? A signature that involves an element and an arbitrary dataframe as input seems hard to explain in a docstring, what is their relation (note that the raster element here doesn't contain a DataFrame, it is its source dataset that does)?

Reply:
Perhaps this should just be:

```python
ds = raster.dataset.clone(df)
return (ds.dimension_values(kd) for kd in raster.kdims)
```

Reply:
That is a nice simplification and I like it. My question then is whether the kdims (default 'x' and 'y') can be set to something else when there is just a 'geometry' column. For instance, does redim work? In general having kdims that aren't actual columns in the dataframe is kinda weird (though I guess there is no alternatives for these geodataframes?).

Reply:
Applied this suggestion in 458508a and it seems to work fine. @philippjfr Just curious about my earlier questions though...I suspect we discussed this before but I can't quite remember (maybe it is a case where the situation is not ideal and not easy to address properly).